docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,225 @@
+# ADR 047: Extend Upgrade Plan
+
+## Changelog
+
+- Nov, 23, 2021: Initial Draft
+
+## Status
+
+PROPOSED Not Implemented
+
+## Abstract
+
+This ADR expands the existing x/upgrade `Plan` proto message to include new fields for defining pre-run and post-run processes within upgrade tooling.
+It also defines a structure for providing downloadable artifacts involved in an upgrade.
+
+## Context
+
+The `upgrade` module in conjunction with Cosmovisor are designed to facilitate and automate a blockchain's transition from one version to another.
+
+Users submit a software upgrade governance proposal containing an upgrade `Plan`.
+The [Plan](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/proto/cosmos/upgrade/v1beta1/upgrade.proto#L12) currently contains the following fields:
+- `name`: A short string identifying the new version.
+- `height`: The chain height at which the upgrade is to be performed.
+- `info`: A string containing information about the upgrade.
+
+The `info` string can be anything.
+However, Cosmovisor will try to use the `info` field to automatically download a new version of the blockchain executable.
+For the auto-download to work, Cosmovisor expects it to be either a stringified JSON object (with a specific structure defined through documentation), or a URL that will return such JSON.
+The JSON object identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
+Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.
+
+If the URL returns an archive, it is decompressed into `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
+Then, if `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}` does not exist, but `{DAEMON_HOME}/cosmovisor/{upgrade name}/{DAEMON_NAME}` does, the latter is copied to the former.
+If the URL returns something other than an archive, it is downloaded to `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`.
+
+If an upgrade height is reached and the new version of the executable version isn't available, Cosmovisor will stop running.
```

Comment:
Yes. There are several reasons this might happen, and Cosmovisor outputs different errors for each before stopping.

Possible reasons:
1. There is no url for the given architecture:
2. The url for the given architecture does not return a file or archive.
3. A checksum was provided for the asset that doesn't match the asset retrieved from the URL.
4. The asset is an archive and a `$DAEMON_NAME` file cannot be found in either `/bin` or `/`.
5. The `/bin/$DAEMON_NAME` file exists but is not executable and cannot be made executable. This might happen if a node operator manually downloads and places the new file without making it executable, and Cosmovisor doesn't have the permissions needed to fix it.